File: analysis/kalman-filter/fOptions.cpp
/***********************************************
 * Deal with options in C++
 *
 * @author Felix Touchte Codjo
 * @date January 17, 2026
 * ********************************************/

#include "fOptions.h"
#include <cstdio>

fOptions::fOptions(std::vector<std::string> _opts) { 
    opts = _opts;
    for (auto x : _opts) {
        mopts[x] = {};
    }
}

bool fOptions::IsOption(std::string str) {
    for (std::string opt: opts) {
        if (opt.compare(str) == 0) {
            return true;
        }
    }
    return false;
}

void fOptions::LoadOptions(int argc, const char * argv[]) {
    // for (int i = 1; i < argc; i++) {
    //     if (IsOption(argv[i]) && i < argc-1) {
    //         mopts[argv[i]] = !IsOption(argv[i+1]) ? argv[i+1] : "";
    //     }
    // }
    std::string current_opt = "default";
    for (int i = 1; i < argc; i++) {
        if (IsOption(argv[i])) {
            current_opt = argv[i];
        } else {
            mopts[current_opt].push_back(argv[i]);
        }
    }
}

std::string fOptions::GetValue(std::string str) {
    if (mopts[str].size() > 0) {
        return mopts[str][0];
    }
    return "";
}

std::vector<std::string> fOptions::GetValues(std::string str) {
    return mopts[str];
}

void fOptions::Show() {
    int max_length = 0;
    for (std::string opt: opts) {
        if ((int) opt.size() > max_length) { 
            max_length = opt.size();
        }
    }
    printf("| List of options\n");
    for (std::string opt: opts) {
        //printf("|   %s = %s\n", opt.c_str(), mopts[opt].c_str());
        printf("|   %s = ", opt.c_str());
        for (auto x : mopts[opt]) {
            printf("%s ", x.c_str());
        }
        printf("\n");
    }
}

File: java-utils/src/main/java/io/github/ftouchte/fOptions.java
package io.github.ftouchte;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

/** Class for managing options in java
 * 
 * java adaptation of amon/analysis/kalman-filter/fOptions.cpp
 * 
 * @author Felix Touchte Codjo
 */
public class fOptions {
    ArrayList<String> opts = new ArrayList<>();
    Map<String, ArrayList<String>> mopts = new HashMap<>();

    /**
     * @brief Construct a new fOptions object
     * 
     * The user is free to give any sense to the list of options. 
     * e.g fOptions("-i", "-o") where -i stands for input and -o for output
     * 
     * @param _opts list of options
     */
    fOptions(String... _opts) {
        for (String x : _opts) {
            opts.add(x);
            mopts.put(x, new ArrayList<>());
        }
    }

    /**
     * @brief Load/parse options
     * 
     * @param args
     */
    void LoadOptions(String[] args) {
        String opt = "default";
        mopts.put(opt, new ArrayList<>());
        for (String x : args) {
            if (IsOption(x)) {
               opt = x;
            }
            else {
                mopts.get(opt).add(x);
            }
        }
    }

    /**
     * @brief Check if str is an option
     * 
     * @param str 
     * @return true str is an option defined at the construction of the object
     * @return false otherwise
     */
    boolean IsOption(String str) {
        for (String x : opts) {
            if (x.equals(str)) {
                return true;
            }
        }
        return false;
    }

    /**
     * @brief Get all values for this option
     * 
     * @param str 
     * @return list of arguments for this option
     */
    ArrayList<String> GetValues(String str) {
        if (IsOption(str)) {
            return new ArrayList<>(mopts.get(str));
        }
        else {
            return null;
        }
        
    }

    /**
     * @brief Get the first value of the option
     * 
     * @param str
     * @return first argument for this option
     */
    String GetValue(String str) {
        if (IsOption(str)) {
            if (mopts.get(str).size() > 0) {
                return mopts.get(str).get(0);
            }
            else {
                return "";
            }
        } else {
            return "";
        }
    }

    /**
     * @brief Display the value for each option
     * 
     */
    void Show() {
        int max_length = 0;
        for (String x : opts) {
            if (x.length() > max_length) {
                max_length = x.length();
            }
        }
        System.out.println("| List of options");
        for (String opt : opts) {
            System.out.printf("|   %-" + max_length + "s = ", opt);
            int counter = 0;
            ArrayList<String> list = GetValues(opt);
            for (String x : list) {
                if (counter >= 10) break;
                System.out.printf("%s ", x);
            }
            if (counter < 10) {
                System.out.println();
            } else {
                System.out.printf("and %d more...\n", list.size() - counter);
            }
        }
    }




}
